tools/deploy/helper.go

```diff
@@ -0,0 +1,370 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/sync"
+)
+
+var (
+	errNoDeploymentProgress = errors.New("no deployment progress")
+)
+
+// DeploymentHelper is helper class handling deployments.
+type DeploymentHelper interface {
+	// Deploy deploys a target revision to the instances in the placement.
+	Deploy(revision string) error
+}
+
+// TODO(xichen): disable deployment while another is ongoing.
+type deploymentHelper struct {
+	logger             xlog.Logger
+	mgr                DeploymentManager
+	planner            DeploymentPlanner
+	client             AggregatorClient
+	store              kv.Store
+	retrier            xretry.Retrier
+	workers            xsync.WorkerPool
+	placementWatcher   services.StagedPlacementWatcher
+	settleBetweenSteps time.Duration
+}
+
+// NewDeploymentHelper creates a new deployment helper.
+func NewDeploymentHelper(opts DeploymentHelperOptions) (DeploymentHelper, error) {
+	placementWatcherOpts := opts.StagedPlacementWatcherOptions()
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
+	if err := placementWatcher.Watch(); err != nil {
+		return nil, err
+	}
+	return deploymentHelper{
+		logger:             opts.InstrumentOptions().Logger(),
+		mgr:                opts.DeploymentManager(),
+		planner:            opts.DeploymentPlanner(),
+		client:             opts.AggregatorClient(),
+		store:              opts.KVStore(),
+		retrier:            opts.Retrier(),
+		workers:            opts.WorkerPool(),
+		placementWatcher:   placementWatcher,
+		settleBetweenSteps: opts.SettleDurationBetweenSteps(),
+	}, nil
+}
+
+func (h deploymentHelper) Deploy(revision string) error {
```

Comment:
Sure I can check whether it's empty or not, don't want to put too many constraints on what `revision` could be or otherwise it's going to compromise the genericity of the deployment tool.